### PR TITLE
[Feat] Add image loading indicator

### PR DESCRIPTION
**Pull request by mdneyazahmad**
### Details

Adds an image loading indicator when images loads in chat and when an image viewer is opened.
### Fixed Issues

$ https://github.com/Expensify/App/issues/7905
$ https://github.com/Expensify/App/issues/7584

### Tests

1. Open any chat
2. Verify that when images loads it as a loading indicator.
3. Click on an image (better if large size)
4. Verify that a loading spinner appears when the image is being downloaded
- [x] Verify that no errors appear in the JS console

#### Contributor (PR Author) Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [ ] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [ ] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. “toggleReport” and not “onIconClick”)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct english, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [x] I verified any copy / text that was added to the app is correct english and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY
- [x] I verified any variables that can be defined as constants (ie. in CONST.js) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] Any functional components have the `displayName` property
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] Any internal methods are bound to “this” properly so there are no scoping issues
    - [x] Any internal methods bound to “this” are necessary to be bound
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose and it is
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn’t already exist
    - [x] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function
(i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)

#### PR Reviewer Checklist
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [x] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. “toggleReport” and not “onIconClick”).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct english, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct english and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components are not impacted by changes in this PR (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY
- [x] I verified any variables that can be defined as constants (ie. in CONST.js) are defined as such
- [x] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] Any internal methods are bound to “this” properly so there are no scoping issues
    - [ ] Any internal methods bound to “this” are necessary to be bound
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose and it is broken down into smaller components in order to separate concerns and functions
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn’t already exist
    - [ ] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function
(i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)

### QA Steps

1. Open any chat
2. Verify that when images loads it as a loading indicator.
3. Click on an image (better if large size)
4. Verify that a loading spinner appears when the image is being downloaded
5. Verify that it does not keeps loading, finally it has to stop.
- [x] Verify that no errors appear in the JS console

#### Web

https://user-images.githubusercontent.com/77761491/159119069-6e537ad1-ab38-45ec-9863-99a6aad8f93a.mp4

#### Mobile Web

https://user-images.githubusercontent.com/77761491/159119114-fdb1ab0d-9080-40c4-aebd-08ed916ed541.mp4

#### Desktop

https://user-images.githubusercontent.com/77761491/159119130-ffc4fa70-782a-433b-8183-458163f83c58.mp4

#### iOS

https://user-images.githubusercontent.com/77761491/159119152-fc76ee27-c104-49a0-9441-068b37b4a3db.mp4

#### Android

https://user-images.githubusercontent.com/77761491/159119168-396ec01f-37b3-4c5b-8769-d3139aaf72ff.mp4

### CODE REVIEW

**Comment by mdneyazahmad** [comment]
This pr is slightly different from my initial proposal. I am working on a very similar issue and found a performance related issue https://github.com/Expensify/App/pull/8203#issuecomment-1072944196 This pr includes that fix.

**Comment by mountiny** [comment]
@mdneyazahmad As mentioned in https://github.com/Expensify/App/pull/8203#issuecomment-1072944196, would you be able to update this PR with the code from the other one? I would revert the previous one as we dont want to push the problem to staging.

Let's make sure it works here thoroughly

**Comment by mountiny** [comment]
I have updated the body of this PR to also sin https://github.com/Expensify/App/issues/7584 issue. Let's update the code with the lines needed from reverted PR and then test it all here. The features are similar (same).

Also dont forget to update the QA and test steps please to reflect that.

Thank you very much!

**Review by mountiny — COMMENTED** [review]
Otherwise, these changes so far look good to me, just as I mentioned, we will need to add the previous PR with the `imageLoadingStart` as class functions.

Thanks!

**Comment by mdneyazahmad** [comment]
@mountiny I will update it with that fix too. Thank you!

**Comment by mountiny** [comment]
Thanks!

@parasharrajat just to clarify, I have asked @mdneyazahmad to include changes from https://github.com/Expensify/App/issues/7584 where we had to revert the PR. These changes are very similar, so it made sense to include them together in one PR which is still manageable and testeable.

**Comment by parasharrajat** [comment]
I am fine with it @mountiny .

**Review by parasharrajat — CHANGES_REQUESTED** [review]
image

Are you following rules from the checklist? Please carefully follow the checklist. I know this could be an extra effort but reviewers also follow the same. Not following it and putting up the PR for review, will indicate a lack of testing. You have already worked on a few issues, we expect you to be more proactive. Thanks.

**Review by parasharrajat — CHANGES_REQUESTED** [review]
There is a lot of duplicate code across components. Let's dry this up. Thanks.

**Comment by mdneyazahmad** [comment]
@parasharrajat I really wanted to discuss the refactoring Image component as it is used at many places. I want to create a wrapper component for `Image` that manages loading indicator internally.

**Comment by parasharrajat** [comment]
Sure, sounds good.

**Comment by mdneyazahmad** [comment]
Updated

**Comment by mdneyazahmad** [comment]
I did not create the wrapper component to manage loading state internally because this component will be used at many places and that will not be much useful we still have to manage the loading state and disable user interaction (modal).

**Review by parasharrajat — CHANGES_REQUESTED** [review]
I can still see one cleanup.

Move the whole loader thing with Image into the component and then just use it everywhere needed. The name could be `ImageWithLoader`.

**Comment by mountiny** [comment]
@parasharrajat seems like your specific comment was not submitted.

**Comment by parasharrajat** [comment]
Correct. Thanks, @mountiny.

**Comment by mdneyazahmad** [comment]
@parasharrajat that does not work on modal there will be a zoom cursor over the loader.

**Comment by parasharrajat** [comment]
Ok, can't we control that?

**Comment by mdneyazahmad** [comment]
Yes we can control but have to manage loading state on that component with a callback method https://github.com/Expensify/App/pull/8238#issuecomment-1076196851

**Comment by mdneyazahmad** [comment]
@parasharrajat The code looks repetitive but we can not create `ImageWithLoader` and use it. Sometime the `LoadingSpinnerOverlay` has to be on different places and sometime we always want to show it (native thumbnail).

```
// ImageView/index.js


<View>
     <Pressable >
           <Image /> // If we create a separate component it will render spinner overlay here 
     </Pressable >
      // We want spinner overlay here
     <LoadingSpinnerOverlay  />
</View>
```

```
// ImageView/index.native.js


<View>
      <Image />
     <LoadingSpinnerOverlay /> // always
</View>

<View>
     <ImageZoom >
           <Image /> // If we create a separate component it will render spinner overlay here 
            <View />
     </ImageZoom >
      // We want spinner overlay here
     <LoadingSpinnerOverlay  />
</View>
```

File: src/components/LoadingSpinnerOverlay.js
Comment (parasharrajat) [inline src/components/LoadingSpinnerOverlay.js:18]: Does this component look similar to `FullScreenLoadingIndicator`? Can we utilize the same?

**Review by parasharrajat — CHANGES_REQUESTED** [review]
changes....

File: src/components/ImageView/index.native.js
Comment (parasharrajat) [inline src/components/ImageView/index.native.js:138]: Was this change necessary?

**Review by parasharrajat — CHANGES_REQUESTED** [review]
Change

**Comment by parasharrajat** [comment]
> The code looks repetitive but we can not create ImageWithLoader and use it.

I am fine with it if @mountiny is.

File: src/components/ImageView/index.native.js
Comment (mdneyazahmad) [reply src/components/ImageView/index.native.js:138]: It was not necessary. I did this change to be consistent.

File: src/components/LoadingSpinnerOverlay.js
Comment (mdneyazahmad) [reply src/components/LoadingSpinnerOverlay.js:18]: It works as expected with `FullScreenLoadingIndicator`.

Comment (parasharrajat) [reply src/components/LoadingSpinnerOverlay.js:18]: So it means, we can't modify it.

**Review by mountiny — APPROVED** [review]
Thank you very much, I have tested it locally and all works well. It all works well so I hope it will not cause any trouble this time!

Thank you @mdneyazahmad was seeing this through, I know this took a long time to polish, but I am glad about your contribution here, thank you very much!

Also thanks @parasharrajat and @rushatgabhane for great reviews as always!

**Comment by mountiny** [comment]
Just asking @parasharrajat if you have anything else you think should be addressed before merging?

**Comment by parasharrajat** [comment]
I think https://github.com/Expensify/App/pull/8238#pullrequestreview-921966220 can be done. 
Apart from it. What about this https://github.com/Expensify/App/issues/7905#issuecomment-1073886714

File: src/components/LoadingSpinnerOverlay.js
Comment (mdneyazahmad) [reply src/components/LoadingSpinnerOverlay.js:18]: I am sorry I missed the `style` props then. We can modify it as `FullScreenLoadingIndicator` accepts a `style` props but the name of this component does not make much sense here. We are showing `FullScreenLoadingIndicator` in place of an `Image`. 

```
<FullscreenLoadingIndicator style={{opacity: 1, backgroundColor: 'transparent'}} />
```

Comment (parasharrajat) [reply src/components/LoadingSpinnerOverlay.js:18]: Yup. Any suggestions for the name. my mind says `FullAreaLoadingIndicator` | `LoadingIndicator` | `CoverLoadingIndicator`

Comment (mdneyazahmad) [reply src/components/LoadingSpinnerOverlay.js:18]: Not sure about it, `FullscreenLoadingIndicator` is a perfect name for what it does. May be we are fine with a separate component or create a generic loader component and create special component with component composition?

Comment (parasharrajat) [reply src/components/LoadingSpinnerOverlay.js:18]: I am sure we can find a suitable name. @mountiny Do you have a suggestion?

**Review by mountiny — CHANGES_REQUESTED** [review]
Thanks!

**Comment by mountiny** [comment]
In regards to https://github.com/Expensify/App/issues/7905#issuecomment-1073886714, I think we should keep that for separate discussion/issue, this one is already dragging for a while.

**Comment by mdneyazahmad** [comment]
> Yeah, better to reuse what we have with the new styles, great catch @parasharrajat!
> 
> ImageLoadingIndicator would sound good to me and it would be clear what it does, could we do that?

I am a bit confused here. `FullScreenLoadingIndicator` can be used in place of image loading indicator and I created `ImageSpinnerOverlay`. Should we rename `FullScreenLoadingIndicator` or `ImageSpinnerOverlay` to `ImageLoadingIndicator`?

**Comment by mountiny** [comment]
@mdneyazahmad Gonna wait for @parasharrajat confirmation, but what I understood is to use the `FullScreenLoadingIndicator` but import it with a nickname `ImageLoadingIndicator`, which would make it clearer in the code what it does. Applying the styles to make it work as expected.

But I am a bit unsure how Rajat meant it in a first-place 😅

**Comment by parasharrajat** [comment]
Yeah, That can do. Good idea. But we usually import the component with the same name. So I was expecting a new name for the component more generic that applies to both locations.

**Comment by mountiny** [comment]
Ah I see, I have looked around and I could not see a place where we would use an alias so let's not do that, you are right.

I am not so sure about renaming the entire component because of this change. The component has many uses its name makes lots of sense so cutting into this would not be so beneficial. Let's just use it as it is since the use case is quite clear, it is a loading indicator, the styling is then explicitly given in the props.

Would you agree? @parasharrajat

**Comment by parasharrajat** [comment]
Yup. I have no problem with that.

**Comment by mountiny** [comment]
Okok, thanks!

@mdneyazahmad finally then last change, could you please reuse the aforementioned component with the stale props to get rid of the opaque?

Thank you snd if it tests well, i will merge then. Really appreciate your patience here.

**Review by mountiny — COMMENTED** [review]
👆

**Comment by mdneyazahmad** [comment]
Any suggestion for the name. We need to create a style for opacity. Is `opacityFull` good name?
```
    opacityFull: {
        opacity: 1,
    },
```

Thanks

**Comment by rushatgabhane** [comment]
@mdneyazahmad `opacity1` is good enough

**Comment by mdneyazahmad** [comment]
Updated

**Comment by mountiny** [comment]
@rushatgabhane My dev machine is a little bit broken after updating my mac to mac os monterey, would you be able to give this a little final test before merging making sure all the images load correctly? I tested before the last changes and all worked fine but just to make sure. Thank you very much!

**Comment by parasharrajat** [comment]
I am also facing the same issue @mountiny.

**Review by parasharrajat — COMMENTED** [review]
Changes look fine.

**Comment by rushatgabhane** [comment]
Sure thing!

**Comment by rushatgabhane** [comment]
Bug: Loading indicator disappears when going from offline to online in attachment view.

1. Go offline.
2. Open an image that isn't cached (opening first time)
We'll see a loading indicator as expected.
3. Go online.
4. The loading indicator disappears! And we have to reopen the attachment view to load the image.


https://user-images.githubusercontent.com/29673073/160888186-2feab5e0-7be4-4dfa-987a-68395d334e7c.mov

**Comment by mountiny** [comment]
@rushatgabhane Great testing!

@mdneyazahmad As this is a new feature, we will try to make sure to test this as much as possible. I realize this is taking longer now, so I would be fine with doubling the reward for one of the linked issues 👍 

Lets get this bug fixed and make sure everything tested well

**Comment by mdneyazahmad** [comment]
I tested with the steps, but the behaviour is different in my case. At step 4 the loading spinner does not go away have to reopen in order to work (may be due to my internet connection is usually slow). The only condition the spinner go away is either `success` or `failed`. In your case it seems to be the latter case and we do not have any logic to handle that case yet. I also see that the production website (without this pr) does not load images at step 4.

> I realize this is taking longer now

I do not see this issue. Could you please confirm it? I create a log in the `imageLoadingStart` and it is calling 1 time as expected. It means, it is downloading it just once.

cc: @rushatgabhane @mountiny

**Comment by parasharrajat** [comment]
> The loading indicator disappears! And we have to reopen the attachment view to load the image.

IMO, this is not part of this PR. @rushatgabhane you should report this on slack. we should show an inline message to indicate that the image load failed.

**Comment by rushatgabhane** [comment]
> we should show an inline message to indicate that the image load failed.

Okay, I'll start a discussion on slack on how to handle failed pessimistic request for an image. I agree that this should be done in a different issue.

**Comment by rushatgabhane** [comment]
I think this is a good opportunity to bring up the alternative approach again. https://github.com/Expensify/App/issues/7905#issuecomment-1073487299

It should work online, offline, and even on slow and flaky connections. What does everyone think here? 😄
Should we consider starting a discussion on the alternative approach if fixing the perpetual spinner is too messy and not worth it?

**Comment by parasharrajat** [comment]
that will not be the case always. We don't always have a thumbnail. And web always needs to fetch the image from the URL. Until we have offline images, we can't do anything. Also, I am totally against showing a low-resolution thumbnail as fullscreen. If you want this, then images should be progressive.

**Comment by rushatgabhane** [comment]
> Also, I am totally against showing a low-resolution thumbnail as fullscreen

Okay, I see that Tim is against this too. 
Alright then, I'm marking my comments as off topic

**Comment by mdneyazahmad** [comment]
> The loader should stop either when the image is loaded or failed to load. Continuous loader is a bug.

On slow connection it takes time to download the image and it does not trigger `onEndLoad`. If it has timeout it would trigger `onError` after `timeoutDuration` and then we display the error or give them a button to reload the image.

**Comment by rushatgabhane** [comment]
@mdneyazahmad sorry I don't understand.
Is there a timeout when downloading images?

**Comment by rushatgabhane** [comment]
> If it has timeout it would trigger onError after timeoutDuration

@mdneyazahmad sorry, I don't understand.
Isn't that the same thing as failed request?

If yes, I think it's best that we handle it in a seperate issue

**Comment by mdneyazahmad** [comment]
@rushatgabhane `Image` component does not have this feature. That was the flow I am suggesting.

**Review by rushatgabhane — APPROVED** [review]
@mountiny LGTM!

The [bug that I mentioned earlier](https://github.com/Expensify/App/pull/8238#issuecomment-1083380917) should be dealt in a different issue. Because it's a part of [offline UX discussion](https://expensify.slack.com/archives/C01GTK53T8Q/p1647628128884459)

~~(I've started a discussion for this [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1648836486717179?thread_ts=1646675711.983089&cid=C01GTK53T8Q))~~

**Comment by mountiny** [comment]
@rushatgabhane @parasharrajat @mdneyazahmad Thank you very much for the discussion here and handling it, I was ooo finishing bachelor thesis 🥵 

I agree that showing low resolution picture on fullscreen is big ⛔ 

And I am happy to sort the slow connection loadings as part of the Offline project pending this PR does not introduce any new problems we havent had before!

**Review by mountiny — APPROVED** [review]

**Comment by mountiny** [comment]
@mdneyazahmad Thank you for your patience on this feature, hopefully, all will work out this time!

**Comment by rushatgabhane** [comment]
> bachelor thesis

@mountiny that makes the two of us 😫

**Comment by mountiny** [comment]
@rushatgabhane You got this!

**Comment by s77rt** [comment]
This PR is partly responsible for a regression here https://github.com/Expensify/App/issues/15288. The use of more than one callback is bugged (https://github.com/DylanVann/react-native-fast-image/issues/975), in this PR it was the use of both `onLoadStart` and `onLoadEnd` where the execution of those two callbacks is not always in order.

**Comment by sobitneupane** [comment]
This PR has probably caused [this regression](https://github.com/Expensify/App/issues/18040).

Issue: Loading spinner is cut off on long images when reopening the chat

**Pull request merged**
